be/src/util/buffered_stream.cpp

```diff
@@ -64,4 +66,87 @@ Status BufferedInputStream::_read_data() {
+    seek_to(offset);
+    return get_bytes(buffer, nbytes, false);
+}
+
+// ===================================================================================
+
+SharedBufferedInputStream::SharedBufferedInputStream(RandomAccessFile* file) : _file(file) {}
+
+Status SharedBufferedInputStream::set_io_ranges(const std::vector<IORange>& ranges) {
+    if (ranges.size() == 0) {
+        return Status::OK();
+    }
+    std::vector<IORange> check(ranges);
+    std::sort(check.begin(), check.end());
+    // check io range is not overlapped.
+    for (size_t i = 1; i < ranges.size(); i++) {
+        if (check[i].offset < (check[i - 1].offset + check[i - 1].size)) {
+            return Status::RuntimeError("io ranges are overalpped");
+        }
+    }
+
+    std::vector<IORange> small_ranges;
+    for (const IORange& r : check) {
+        if (r.size > _options.max_buffer_size) {
+            SharedBuffer sb = SharedBuffer{.offset = r.offset, .size = r.size, .ref_count = 1};
+            _map.insert(std::make_pair(r.offset + r.size, sb));
+        } else {
+            small_ranges.emplace_back(r);
+        }
+    }
+
+    if (small_ranges.size() > 0) {
+        auto update_map = [&](size_t from, size_t to) {
+            // merge from [unmerge, i-1]
+            int64_t ref_count = (to - from + 1);
+            int64_t end = (small_ranges[to].offset + small_ranges[to].size);
+            SharedBuffer sb = SharedBuffer{.offset = small_ranges[from].offset,
+                                           .size = end - small_ranges[from].offset,
+                                           .ref_count = ref_count};
+            _map.insert(std::make_pair(sb.offset + sb.size, sb));
+        };
+
+        size_t unmerge = 0;
+        for (size_t i = 1; i < small_ranges.size(); i++) {
+            const auto& prev = small_ranges[i - 1];
+            const auto& now = small_ranges[i];
+            size_t now_end = now.offset + now.size;
+            size_t prev_end = prev.offset + prev.size;
+            if (((now_end - small_ranges[unmerge].offset) <= _options.max_read_size) &&
+                (now.offset - prev_end) <= _options.max_dist_size) {
+                continue;
+            } else {
+                update_map(unmerge, i - 1);
+                unmerge = i;
+            }
+        }
+        update_map(unmerge, small_ranges.size() - 1);
+    }
+    return Status::OK();
+}
+
+Status SharedBufferedInputStream::get_bytes(const uint8_t** buffer, size_t offset, size_t* nbytes) {
+    auto iter = _map.upper_bound(offset);
+    if (iter == _map.end()) {
+        return Status::RuntimeError("failed to find shared buffer based on offset");
+    }
+    SharedBuffer& sb = iter->second;
+    if ((sb.offset > offset) || (sb.offset + sb.size) < (offset + *nbytes)) {
+        return Status::RuntimeError("bad construction of shared buffer");
+    }
+    if (sb.buffer.capacity() == 0) {
+        sb.buffer.reserve(sb.size);
```

Comment:
Is there any memory restriction?

Reply:
Yes it will consume large memory, for a column, the layout is 

```
| page header1 | data page1 | page header2 | data page2 |
```

- the old way of reading is to: 
  - read page head1
  - read data page1
  - read PH2, read DP2
- the new way of reading is to:
  - read page header1 to data page2 in a shot.

The first way, max memory is `max(ph1, dp1, ph2, dp2)`. The second way is `ph1 + dp1 + ph2 + dp2`.